Stop using Python's eval() for -m and -k

bluetech opened the pull request:
Previously, the expressions given to the `-m` and `-k` options were evaluated with `eval`. This causes a few issues:

- Python keywords cannot be used.

- Constants like numbers, None, True, False are not handled correctly.

- Various syntax like numeric operators and `X if Y else Z` is supported
  unintentionally.

- `eval()` is somewhat dangerous for arbitrary input.

- Can fail in many ways so requires `except Exception`.

The format we want to support is quite simple, so change to a custom parser. This fixes the issues above, and gives us full control of the format, so can be documented comprehensively and even be extended in the future if we wish.

Fixes #1141.
Fixes #3573.
Fixes #5881.
Fixes #6822.
Fixes #7112.

<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ ] Add yourself to `AUTHORS` in alphabetical order.
-->

RonnyPfannschmidt commented:
Great work, should go into a major release as breaking change

nicoddemus commented:
Hey @RonnyPfannschmidt,

Hmm what does it break? Are `MarkMappings` and `KeywordMappings` accessible by the users?

RonnyPfannschmidt commented:
Changing the behaviour constitutes a breaking change of the cli api for the arguments

nicoddemus commented:
> Changing the behaviour constitutes a breaking change of the cli api for the arguments

Definitely, but IIUC this is not changing the behavior, it is only making it more stable/safe? Can you show an example of a command-line that was possible before and is not anymore?

RonnyPfannschmidt commented:
Exclusive or expression for marks/keywords

nicoddemus commented:
Ahh indeed:

```
 λ pytest testing\test_compat.py -k "safe ^ _get" -v
======================== test session starts ========================
collected 11 items / 8 deselected / 3 selected

testing/test_compat.py::test_get_real_func PASSED              [ 33%]
testing/test_compat.py::test_get_real_func_partial PASSED      [ 66%]
testing/test_compat.py::test_safe_isclass PASSED               [100%]

================== 3 passed, 8 deselected in 0.04s ==================
```

Also many others are probably possible: `safe + _get`.

How should we proceed then:

1. Keep this unmerged until we deem we should release 6.0?
2. Merge and make the next release 6.0, regardless of other features that might be on the queue for the next major? Here's what we have on the milestone: https://github.com/pytest-dev/pytest/milestone/24

RonnyPfannschmidt commented:
Not sure, we might want to defer part of 6.0 to 7.0, im not in a position to get into it in deeply

nicoddemus commented:
Sounds reasonable!

We might release 5.4.2 first, then follow up with 6.0.0 from `master`.

For 6.0.0 I think https://github.com/pytest-dev/pytest/issues/5584 is a must have, and https://github.com/pytest-dev/pytest/issues/6324 is nice to have. What do you think, @RonnyPfannschmidt and @bluetech?

bluetech commented on src/_pytest/mark/legacy.py:
```diff
@@ -11,35 +16,174 @@
+                yield Token(TokenType.RPAREN, ")", pos + 1)
+                pos += 1
+            else:
+                match = re.match(r"(:?\w|\+|-|\.|\[|\])+", input[pos:])
```
Yes, that'd be good. (I'll apply this a bit latter with a test case).

bluetech commented:
> Exclusive or expression for marks/keywords

This (and similar) can be added if we want to support them. As for `^` specifically, for bools it feels a little weird. If Python had a `xor` operator that would have been nice :)

> We might release 5.4.2 first, then follow up with 6.0.0 from master.

:+1:

It will be good to do a 5.4.2. There is some unittest issue which many people seem to complain about but I haven't looked into the details myself.

For now I'll mark this PR as "draft" until we decide what to exactly.

For 6.0.0, since it's a major release, as per the issue I think we should audit the deprecations and decide which we can remove.

Will also be nice to have some highlight features -- I obviously would like to see the type annotations go in, but it's a big branch which is not so easy for straight review.

bluetech commented:
@nicoddemus 


> Perhaps move the parsing stuff to its own module, including evaluate and up, but leaving only the pytest specific code here, like MarkMatcher, KeywordMatcher, etc?

That makes sense.

Related to this, one thing I'm curious about is the name `_pytest.mark.legacy` for the current module. I don't really see anything legacy in there. Is this name just some legacy (sorry for the pun :smile:)? If so, would be nice to rename it, though that would break any external code which imports from it.

> Some unittests for evaluate would be nice too, I think.

Will do.

RonnyPfannschmidt commented:
originally i wanted to just deprecate and drop mark and keyword expressions as they are

mark expressions can't select marks based on mark configuration

keyword expressions are a most confusion concept as they throw test id components, random paths and random attributes of the objects to be tested into the keyword listing

i was hoping to have some preexisting xpath/jq style language to select tests based on the data they expose

Sup3rGeo commented:
> This (and similar) can be added if we want to support them. As for `^` specifically, for bools it feels a little weird. If Python had a `xor` operator that would have been nice :)

My thoughts exactly - can't we add the `^` also to this and make this non-breaking, and available in a minor release?

Also, because we are rolling our own thing, we could also add `xor` even if python does not have it a operator?

RonnyPfannschmidt commented:
another detail is that python operators behave dfferently, xor doesnt make sense for a shortcut operator - the non-shortcut operators for "propper boolean logic" are the special character ones

nicoddemus commented:
> My thoughts exactly - can't we add the ^ also to this and make this non-breaking, and available in a minor release?

`^` is only one example... because previously we supported `eval()`, any Python expression technically works, even `safe + _get` for example. So it is technically a backward incompatible change.

But fear not, I believe we should release 6.0.0 soonish, we just need to release 5.4.2 first (probably solving the existing unittest issue).

bluetech commented:
Updated:
- Split to separate file `mark/expression.py`
- Added unit tests
- Allowed `:` in idents
- Allowed empty expression, evaluates to false

I marked this as ready now, with a 6.0 milestone.

bluetech commented:
So given that `master` is not definitely 6.0, I am going to merge this. Thanks everyone for review/discussion.

It's probable we'll get issues of expressions that stopped working. But I think we should evaluate extensions to the syntax on a case-by-case basis.

asottile commented:
@bluetech I'm seeing some odd failures in #7179 that look related -- wondering if you could take a look at that:

```
2020-05-07T12:32:22.0720900Z ______________________ test_keyword_option_custom[spec5] _______________________
2020-05-07T12:32:22.0722120Z [gw3] darwin -- Python 3.7.7 /Users/runner/runners/2.169.1/work/pytest/pytest/.tox/py37-xdist/bin/python
2020-05-07T12:32:22.0722520Z 
2020-05-07T12:32:22.0723240Z spec = ('1 or 2', ('test_1', 'test_2'))
2020-05-07T12:32:22.0724860Z testdir = <Testdir local('/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/popen-gw3/test_keyword_option_custom1')>
2020-05-07T12:32:22.0725220Z 
2020-05-07T12:32:22.0726620Z     @pytest.mark.parametrize(
2020-05-07T12:32:22.0726800Z         "spec",
2020-05-07T12:32:22.0726920Z         [
2020-05-07T12:32:22.0727060Z             ("interface", ("test_interface",)),
2020-05-07T12:32:22.0727190Z             ("not interface", ("test_nointer", "test_pass", "test_1", "test_2")),
2020-05-07T12:32:22.0734390Z             ("pass", ("test_pass",)),
2020-05-07T12:32:22.0734620Z             ("not pass", ("test_interface", "test_nointer", "test_1", "test_2")),
2020-05-07T12:32:22.0734790Z             ("not not not (pass)", ("test_interface", "test_nointer", "test_1", "test_2")),
2020-05-07T12:32:22.0734990Z             ("1 or 2", ("test_1", "test_2")),
2020-05-07T12:32:22.0735790Z         ],
2020-05-07T12:32:22.0735950Z     )
2020-05-07T12:32:22.0736720Z     def test_keyword_option_custom(spec, testdir):
2020-05-07T12:32:22.0736900Z         testdir.makepyfile(
2020-05-07T12:32:22.0737020Z             """
2020-05-07T12:32:22.0737110Z             def test_interface():
2020-05-07T12:32:22.0737230Z                 pass
2020-05-07T12:32:22.0737340Z             def test_nointer():
2020-05-07T12:32:22.0737460Z                 pass
2020-05-07T12:32:22.0737540Z             def test_pass():
2020-05-07T12:32:22.0737780Z                 pass
2020-05-07T12:32:22.0737880Z             def test_1():
2020-05-07T12:32:22.0737970Z                 pass
2020-05-07T12:32:22.0738050Z             def test_2():
2020-05-07T12:32:22.0738170Z                 pass
2020-05-07T12:32:22.0738260Z         """
2020-05-07T12:32:22.0738360Z         )
2020-05-07T12:32:22.0738470Z         opt, passed_result = spec
2020-05-07T12:32:22.0740700Z         rec = testdir.inline_run("-k", opt)
2020-05-07T12:32:22.0740870Z         passed, skipped, fail = rec.listoutcomes()
2020-05-07T12:32:22.0741910Z         passed = [x.nodeid.split("::")[-1] for x in passed]
2020-05-07T12:32:22.0742070Z >       assert len(passed) == len(passed_result)
2020-05-07T12:32:22.0742150Z E       AssertionError: assert 5 == 2
2020-05-07T12:32:22.0742640Z E        +  where 5 = len(['test_interface', 'test_nointer', 'test_pass', 'test_1', 'test_2'])
2020-05-07T12:32:22.0743090Z E        +  and   2 = len(('test_1', 'test_2'))
2020-05-07T12:32:22.0743150Z 
2020-05-07T12:32:22.0743280Z /Users/runner/runners/2.169.1/work/pytest/pytest/testing/test_mark.py:289: AssertionError
```

bluetech commented:
That's an odd one, I'll try to reproduce it in #7184.

The-Compiler commented:
@bluetech I found it quite interesting how you implemented this parser, using functions for the different parts of the grammar. Where did you get that idea from? Do you know of some write-up/book/... explaining that technique? This is the first time I've ever seen it, and I might need to write a simple parser myself soon-ish.

bluetech commented:
The keywords would be "recursive descent parser", "predictive parser" or "top-down parser".

As long as the grammar is simple and non-left-recursive, implementing it with recursive functions is probably the most straightforward way to go about it.
